owncloudComLibrary/src/main/java/com/owncloud/android/lib/common/http/methods/webdav/DavMethod.kt

```diff
@@ -0,0 +1,164 @@
+import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
+import okhttp3.Protocol
+import okhttp3.Response
+import okhttp3.ResponseBody.Companion.toResponseBody
+import java.net.MalformedURLException
+import java.net.URL
+import java.util.concurrent.TimeUnit
+
+/**
+ * Wrapper to perform WebDAV (dav4android) calls
+ *
+ * @author David González Verdugo
+ */
+abstract class DavMethod protected constructor(url: URL) : HttpBaseMethod(url) {
+    protected var mDavResource: DavOCResource
+
+    init {
+        val httpUrl = url.toHttpUrlOrNull() ?: throw MalformedURLException()
+        mDavResource = DavOCResource(
+            okHttpClient,
+            httpUrl,
+            log
+        )
+    }
+
+    override fun abort() {
+        mDavResource.cancelCall()
+    }
+
+    @Throws(Exception::class)
+    override fun execute(): Int {
+        return try {
+            onExecute()
+        } catch (httpException: HttpException) {
+            // Modify responses with information gathered from exceptions
+            if (httpException is RedirectException) {
+                response = Response.Builder()
+                    .header(
+                        HttpConstants.LOCATION_HEADER, httpException.redirectLocation
+                    )
+                    .code(httpException.code)
+                    .request(request)
+                    .message(httpException.message ?: "")
+                    .protocol(Protocol.HTTP_1_1)
+                    .build()
+            } else {
+                // The check below should be included in okhttp library, method ResponseBody.create(
+                // TODO check most recent versions of okhttp to see if this is already fixed and try to update if so
+                if (response.body?.contentType() != null) {
+                    val responseBody = (httpException.responseBody ?: ""
+                            ).toResponseBody(response.body?.contentType())
+                    response = response.newBuilder()
+                        .body(responseBody)
+                        .build()
+                }
+            }
+            httpException.code
+        }
+    }
+
+    //////////////////////////////
+    //         Setter
+    //////////////////////////////
+    // Connection parameters
+    override fun setReadTimeout(readTimeout: Long, timeUnit: TimeUnit) {
+        super.setReadTimeout(readTimeout, timeUnit)
+        mDavResource = DavOCResource(
+            okHttpClient,
+            httpUrl,
+            log
+        )
+    }
+
+    override fun setConnectionTimeout(
+        connectionTimeout: Long,
+        timeUnit: TimeUnit
+    ) {
+        super.setConnectionTimeout(connectionTimeout, timeUnit)
+        mDavResource = DavOCResource(
+            okHttpClient,
+            httpUrl,
+            log
+        )
+    }
+
+    override fun setFollowRedirects(followRedirects: Boolean) {
+        super.setFollowRedirects(followRedirects)
+        mDavResource = DavOCResource(
```

Comment:
I do not particularly like this approach of creating new `DavOCResources`, I know that it's probably my code, I had a bad day I guess 😄  . 
What I mean is that we are updating the client via `super.setFollowRedirects(followRedirects)` but if you don't go there you won't see it and you won't know why we need a new DavOCResource here...